socs/agents/hwp_pcu/agent.py

```diff
@@ -116,20 +117,32 @@ def main(self, session, params):
                     self.log.info('Connected to PCU')
                     PCU.clear_buffer()
                     self.log.info('Cleared buffer')
-                except ConnectionRefusedError:
+                except (ConnectionRefusedError, serial.serialutil.SerialException):
                     self.log.error(
                         "Could not connect to PCU. "
                         "Retrying after 30 sec..."
                     )
                     time.sleep(30)
                     continue
-            now = time.time()
-            if now - last_daq > 5:
-                self._get_and_publish_data(PCU, session)
-                last_daq = now
-
-            self._process_actions(PCU)
-            time.sleep(0.1)
+            try:
+                now = time.time()
+                if now - last_daq > 5:
+                    self._get_and_publish_data(PCU, session)
+                    last_daq = now
+
+                self._process_actions(PCU)
+                session.degraded = False
+                time.sleep(0.1)
+            except serial.serialutil.SerialException:
+                self.log.error(
+                    "Decive reports readiness to read but returned no data. "
+                    "Reconnect after 30 sec..."
+                )
+                PCU.close()
+                PCU = None
+                session.degraded = True
+                time.sleep(30)
```

Comment:
There's already a 30 second sleep on the reconnect logic in line 124/125, so I'd say you can drop this here. If it can reconnect it'll do so quickly, if it can't it'll wait 30 seconds already.